### PR TITLE
Parallel prime generation

### PR DESCRIPTION
**Pull request by dvdplm**
This adds a new function to the public API, named "par_generate_prime_with_rng", that parallelizes the search for primes.

Outstanding tasks:

- [x] Add the corresponding "par_generate_safe_prime_with_rng"
- [x] ~~Pick the number of threads to use~~
- [x] Decide is it is useful for end users to be able to set the number of threads or not (and if yes, what the API for the should be: generics? another function arg? Env variable?) (7491a06939d96d1e77f9a352bd51fe86f19183b8 adds another function arg)
- [ ] Consider a name change (deferred to #62)
- [x] More and better tests
- [x] `CHANGELOG` entry and better docs
- [x] Write criterion benchmarks

### Results

During this work I have tried out several different parallel implementations, using `rayon` and `chili`. The latter is a new take on low-overhead parallelism and is attractive for its small code size and advertised performance, but in my tests I have not seen good performance from it (see below).

When searching for random primes we do not know how many iterations it will take, only that we will eventually find one. This is reflected in the benchmark results as a big variability, as seen in the "fastest" and "slowest" columns below.

Unsurprisingly, the larger the prime we're looking for, the more we benefit from parallelizing the work. For `Uint<64>`s (aka `U4096`), using 8 CPUs is 6x faster; for `Uint<16>` the speedup is more like 3x.

The benchmarks use the `divan` benchmarking library and ran on a MacBook Pro M3 Max, using 8 threads. The table compares different implementations and sizes. The implementation in this PR is labeled "rayon_find_any2":

```
benches                    fastest       │ slowest       │ median        │ mean          │ samples │ iters
╰─ benches                               │               │               │               │         │
   ├─ pargen                             │               │               │               │         │
   │  ├─ chili_join2                     │               │               │               │         │
   │  │  ├─ Uint<16>       3.248 ms      │ 54.54 ms      │ 11.3 ms       │ 13.97 ms      │ 250     │ 250
   │  │  ├─ Uint<32>       24.23 ms      │ 3.385 s       │ 124.6 ms      │ 189.2 ms      │ 250     │ 250
   │  │  ╰─ Uint<64>       209.8 ms      │ 33.87 s       │ 1.622 s       │ 2.737 s       │ 250     │ 250
   │  ├─ rayon_find_any                  │               │               │               │         │
   │  │  ├─ Uint<16>       2.885 ms      │ 17.25 ms      │ 5.252 ms      │ 5.946 ms      │ 250     │ 250
   │  │  ├─ Uint<32>       20.37 ms      │ 270.7 ms      │ 41.78 ms      │ 56.22 ms      │ 250     │ 250
   │  │  ╰─ Uint<64>       156.7 ms      │ 3.187 s       │ 558.3 ms      │ 683.7 ms      │ 250     │ 250
   │  ├─ rayon_find_any2 <–– THIS PR     │               │               │               │         │ 
   │  │  ├─ Uint<16>       2.825 ms      │ 14.61 ms      │ 5.367 ms      │ 5.742 ms      │ 250     │ 250
   │  │  ├─ Uint<32>       20.23 ms      │ 175.4 ms      │ 46.5 ms       │ 56.24 ms      │ 250     │ 250
   │  │  ╰─ Uint<64>       156.6 ms      │ 2.749 s       │ 523 ms        │ 703.5 ms      │ 250     │ 250
   │  ├─ rayon_join                      │               │               │               │         │
   │  │  ├─ Uint<16>       4.125 ms      │ 16.74 ms      │ 5.569 ms      │ 6.353 ms      │ 250     │ 250
   │  │  ├─ Uint<32>       30.1 ms       │ 256.4 ms      │ 41.14 ms      │ 55.48 ms      │ 250     │ 250
   │  │  ╰─ Uint<64>       233.8 ms      │ 3.044 s       │ 608.6 ms      │ 693.1 ms      │ 250     │ 250
   │  ╰─ single  <–– MASTER              │               │               │               │         │
   │     ├─ Uint<16>       2.276 ms      │ 91.08 ms      │ 14.47 ms      │ 18.83 ms      │ 250     │ 250
   │     ├─ Uint<32>       17.33 ms      │ 1.277 s       │ 180.3 ms      │ 256.4 ms      │ 250     │ 250
   │     ╰─ Uint<64>       156.2 ms      │ 18.13 s       │ 3.005 s       │ 4.071 s       │ 250     │ 250
```

There is likely ways to do better than what is presented here, but this is a library and not a prime searching application and there's a value in keeping things simple&nimble. A more efficient parallelizing solution would probably use more code to synchronize threads (channels or shared memory); indeed the reason why `chili` is slower than `rayon` here is because the latter has more facilities in place to halt jobs when a prime is found.

### CODE REVIEW

File: src/presets.rs [hunk 1]
```diff
@@ -90,6 +93,45 @@ pub fn generate_safe_prime_with_rng<T: Integer + RandomBits + RandomMod>(
...
+
+    threadpool.install(|| {
+        match sieve.par_bridge().find_any(|c| {
+            //TODO(dp): This clone feels dumb. OTOH it is fairly cheap to clone a rng.
+            let mut rng = rng.clone();
+            is_prime_with_rng(&mut rng, c)
+        }) {
+            // TODO(dp): This clone is very annoying.
+            Some(p) => p.clone(),
+            None => par_generate_prime_with_rng(rng, bit_length.get()),
```
Comment (fjarri) [inline src/presets.rs:130]: Do you hit this line in your tests? I wonder how does creating a thread pool within a thread pool work.

File: src/presets.rs [hunk 2]
```diff
@@ -90,6 +93,45 @@ pub fn generate_safe_prime_with_rng<T: Integer + RandomBits + RandomMod>(
...
+        .num_threads(threadcount)
+        .build()
+        .expect("If the platform can spawn threads, then this call will work.");
+    let start = random_odd_integer::<T>(rng, bit_length).get();
+    let sieve = Sieve::new(start, bit_length, false);
+
+    threadpool.install(|| {
+        match sieve.par_bridge().find_any(|c| {
+            //TODO(dp): This clone feels dumb. OTOH it is fairly cheap to clone a rng.
+            let mut rng = rng.clone();
```
Comment (fjarri) [inline src/presets.rs:125]: In a similar construction in a Monte-Carlo calculation you'd want some kind of a split() method to produce a new instance of RNG deterministically. But I think cloning is probably good enough for prime testing specifically.

**Comment by fjarri** [comment]
5x speedup for 8 threads is a pretty good result. I wonder how will safe prime generation behave. 

Did you try using several `Sieve` objects in any of the other implementations? That would require synchronization between threads though, so that you could cancel them as soon as you find something.

File: src/presets.rs (hunk 1 above)
Comment (dvdplm) [reply src/presets.rs:130]: It happens. Not often, but for 8192bit primes it happens regularly. If we decide to go ahead with this feature I'll make sure to investigate further how/when the threadpool drops.

Comment (dvdplm) [reply src/presets.rs]: So this was a good catch: the threads are not dropped when recursing. I reshuffled things a bit so I can do an explicit `drop` on the threadpool and checked that it works.

File: src/presets.rs (hunk 2 above)
Comment (dvdplm) [reply src/presets.rs:124]: > you'd want some kind of a split() method to produce a new instance of RNG deterministically

Looks like this is indeed a feature of some rust RNGs, notably ChaCha and Xoshiro. The "blackbox" `OsRng` however has nothing of the kind. Cloning will have to due!

**Comment by dvdplm** [comment]
> I wonder how will safe prime generation behave.

Finally got around to benchmarking this yesterday:

```
benches                  fastest       │ slowest       │ median        │ mean          │ samples │ iters
╰─ benches                             │               │               │               │         │
   ╰─ pargen                           │               │               │               │         │
      ╰─ safe_primes                   │               │               │               │         │
         ├─ rayon                      │               │               │               │         │
         │  ├─ Uint<16>  12.21 ms      │ 162 ms        │ 74.58 ms      │ 74.78 ms      │ 13      │ 13
         │  ├─ Uint<32>  319 ms        │ 5.105 s       │ 1.97 s        │ 2.297 s       │ 10      │ 10
         │  ╰─ Uint<64>  11.53 s       │ 3.058 m       │ 1.371 m       │ 1.533 m       │ 10      │ 10
         ╰─ single                     │               │               │               │         │
            ├─ Uint<16>  35.51 ms      │ 2.04 s        │ 530.9 ms      │ 699.2 ms      │ 10      │ 10
            ├─ Uint<32>  762 ms        │ 2.54 m        │ 15.73 s       │ 29.9 s        │ 10      │ 10
            ╰─ Uint<64>  10.26 s       │ 45.91 m       │ 8.388 m       │ 14.11 m       │ 10      │ 10
```

The number of samples is much lower in this run because of how long it takes to find big safe primes, and the code uses all available cores (as opposed to the results in the PR description that uses half the available cores).  The median times are between 6 and 8 times faster using 16 cores and there seems to be an even bigger difference for the slowest/fastest cases (but the sample size is perhaps too small to really tell).

I think this is pretty much what I expected. Thoughts?

File: Cargo.toml [hunk 3]
```diff
@@ -35,6 +39,7 @@ tests-gmp = ["rug/std"]
 tests-glass-pumpkin = ["glass_pumpkin"]
 tests-exhaustive = []
 tests-all = ["tests-openssl", "tests-gmp", "tests-exhaustive", "tests-glass-pumpkin"]
+rayon = ["dep:rayon"]
```
Comment (fjarri) [inline Cargo.toml:42]: Do you think we should expose the specifics of how we do parallelization, or just call the feature `multicore` or something? Does the choice of the thread pool affect the user?

File: Cargo.toml [hunk 4]
```diff
@@ -44,3 +49,8 @@ rustdoc-args = ["--cfg", "docsrs"]
 [[bench]]
 name = "bench"
 harness = false
+
+[profile.bench]
```
Comment (fjarri) [inline Cargo.toml:53]: Was there a specific reason for this? `opt-level=3` I think is the default one, but you're overriding `lto` and `codegen-units`.

File: README.md [hunk 5]
```diff
@@ -16,6 +16,39 @@ In particular:
 
 See the documentation for the specific tests for more information and references.
```
Comment (fjarri) [inline README.md:17]: While you're at it, could you fix this sentence too? I think "for the specific tests" should be removed.

Comment (fjarri) [reply README.md:17]: And perhaps, given the info you added, this sentence can be removed completely.

File: README.md [hunk 6]
```diff
@@ -16,6 +16,39 @@ In particular:
...
+
+```rust
+use crypto_bigint::U1024;
+let prime = crypto_primes::generate_safe_prime::<U1024>(64);
+assert!(crypto_primes::is_safe_prime(&prime));
+```
+
+## Advanced
+
+Advanced users can use the [`hazmat`][hazmat-lnk] module in the library to build a custom prime finding solution that best fit their needs, e.g. by picking different Lucas bases or running Miller-Rabin tests with particular bases.
```
Comment (fjarri) [inline README.md:43]: The link won't work in the README rendered in github, will it? Maybe just use a hard link to `docs.rs` instead?

Comment (fjarri) [reply README.md:43]: Although this way the link will be incorrect when viewing pre-release versions... Not sure what's the good way of dealing with it, except for just putting it all into `lib.rs`.

**Comment by fjarri** [comment]
I think we can deal with #62 in a subsequent PR

File: Cargo.toml (hunk 4 above)
Comment (dvdplm) [reply Cargo.toml:53]: It is indeed the default, I think I just copy&pasted the whole thing from somewhere. Removing.

File: Cargo.toml (hunk 3 above)
Comment (dvdplm) [reply Cargo.toml:42]: I like "multicore"!

File: README.md (hunk 6 above)
Comment (dvdplm) [reply README.md:43]: It'll show the "latest" published (v0.5.0). I think it's ok and that we can probably release 0.6 soon enough that it doesn't matter so much.

**Pull request merged**